diff --git a/objects/window.c b/objects/window.c
index 02e5ace..2610d7c 100644
--- a/objects/window.c
+++ b/objects/window.c
@@ -24,6 +24,7 @@
 #include "ewmh.h"
 #include "screen.h"
 #include "objects/window.h"
+#include "objects/client.h"
 #include "common/atoms.h"
 #include "common/luaobject.h"
 
@@ -167,9 +168,11 @@ luaA_window_set_border_color(lua_State *L, window_t *window)
 void
 window_set_border_width(lua_State *L, int idx, int width)
 {
+    client_t *c;
     window_t *window = luaA_checkudata(L, idx, &window_class);
 
-    if(width == window->border_width || width < 0)
+    if(width == window->border_width || width < 0 ||
+       ((c = client_getbywin(window->window)) && c->fullscreen && width > 0))
         return;
 
     if(window->window)
