diff --git a/luaa.c b/luaa.c
index 2483c80..3d5104b 100644
--- a/luaa.c
+++ b/luaa.c
@@ -271,9 +271,8 @@ luaA_fixups(lua_State *L)
     lua_setfield(L, -2, "wlen");
     lua_pop(L, 1);
     /* replace next */
-    lua_pushliteral(L, "next");
     lua_pushcfunction(L, luaAe_next);
-    lua_settable(L, LUA_GLOBALSINDEX);
+    lua_setglobal(L, "next");
     /* replace pairs */
     lua_pushliteral(L, "pairs");
     lua_pushcfunction(L, luaAe_next);
@@ -285,13 +284,11 @@ luaA_fixups(lua_State *L)
     lua_pushcclosure(L, luaAe_ipairs, 1);
     lua_settable(L, LUA_GLOBALSINDEX);
     /* replace type */
-    lua_pushliteral(L, "type");
     lua_pushcfunction(L, luaAe_type);
-    lua_settable(L, LUA_GLOBALSINDEX);
+    lua_setglobal(L, "type");
     /* set selection */
-    lua_pushliteral(L, "selection");
     lua_pushcfunction(L, luaA_selection_get);
-    lua_settable(L, LUA_GLOBALSINDEX);
+    lua_settable(L, "selection");
 }
 
 /** Look for an item: table, function, etc.
